pkg/scheduler/cache/node_tree.go

```diff
@@ -0,0 +1,183 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	utilnode "k8s.io/kubernetes/pkg/util/node"
+
+	"github.com/golang/glog"
+)
+
+// NodeTree is a tree-like data structure that holds node names in each zone. Zone names are
```

Comment:
I think this entire file needs logging or else it may be difficult to debug.

Reply:
Logging could help, but probably not much for debugging a structure like this. We should rely on testing more. I add some logs as well.

Reply:
True but as a cluster operator, when I am debugging, my first inclination is to increasing debugging level and then look at them to see why a node has been chosen instead of another. Looking at the code or test cases may be my last resort assuming I am familiar with the code base.

Reply:
This is a good point. I am adding a couple of log messages with high verbosity level.